### ISSUE TITLE
MailChimp Firebase Integration only adds email and not first name and last name

### ISSUE THREAD

**Issue by zkhalapyan** [firebase/extensions#411]
### [READ] Step 1: Are you in the right place?

Issues filed here should be about a feature request for a **specific extension in this repository**. To file a feature request that affects multiple extensions or the Firebase Extensions platform, please reach out to
[Firebase support](https://firebase.google.com/support/troubleshooter/report/features/) directly.

### [REQUIRED] Step 2: Extension name

This feature request is for extension: auth-mailchimp-sync

### What feature would you like to see?

Mailchimp audience list should additionally include first name and last name; I am already collecting that information in my list, can you please provide a way to inject that into mailchimp? I think using mailchimp without using user's first and last name is kind of silly and drops conversion even further! 

### How would you use it?

When I send an email campaign to my audience, I would refer to them with their first name, improving the quality of the email.

**Comment by Ehesp** [comment]
Hello! Looking at the docs, it doesn't seem it's possible to add a name field to a member: https://mailchimp.com/developer/api/marketing/list-members/add-or-update-list-member

Unless, it can be provided to `merge_fields`.

**Comment by zkhalapyan** [comment]
Thanks for getting back so quickly @Ehesp ! 

TL;DR; We can use `FNAME` and `LNAME` inside `merge_fields` as documented [here](https://mailchimp.com/developer/guides/create-your-first-audience#add-a-contact-to-an-audience).

Interestingly, their API introduction section does document how add first name and last name fields to the creation request, see [Add a contact to an audience]( https://mailchimp.com/developer/guides/create-your-first-audience#add-a-contact-to-an-audience).

It does seem like they are using merge_fields property in this case with properties `FNAME` and `LNAME`. Unfortunately, their [Merge Fields API documentation](https://mailchimp.com/developer/api/marketing/list-merges) doesn't include all the possible fields, but they do provide some context on it in their [merge field cheat sheet](https://mailchimp.com/help/all-the-merge-tags-cheat-sheet/). 

So their documentation could definitely use some improvement, but I do think this is possible.

**Comment by Ehesp** [comment]
@zkhalapyan The main issue here is that only a "display name" is returned, which isn't specifically a full name. We could perhaps add this to the `FNAME` field if it exists, but it may not be a first name, or may include the last name.

**Comment by zkhalapyan** [comment]
> The main issue here is that only a "display name" is returned.

This is from internal firebase auth? 

I guess there are three options: 

1. Update firebase auth to provide proper first name, last name, display name, etc. structure to get rid of this limitation 
2. Provide simple instructions on how to extend the extension to add in merge_fields (maybe some tutorial steps & code examples catering towards the top most important fields).
3. Push display name as a custom merge_field i.e. DNAME or display_name to mailchimp; at least this gives mailchimp users some resemblance of a name they can use to address their audience in the name. 

Ideally, we could do #1. This is beneficial from many different aspects. If not, I would pick #2. Probably the worst is #3 since it could end up putting in full names inside display names and just give noisy data to mailchimp users. 

WDYT?

**Comment by zkhalapyan** [comment]
Another quick note here... so once a member is created, it is not possible to update the member's name just via the email address. [Updating the name would require knowing the member's ID on mailchimp](https://mailchimp.com/developer/api/marketing/list-members/add-or-update-list-member/). This is possibly, but annoying to go through all the steps.

In other words, by integrating with Firebase's mailchimp extension, you are making it extremely difficult for developers to later update the contacts with names.

**Comment by sam-jg** [comment]
@Ehesp and @zkhalapyan could we just split the display name on spaces, then use the first element as `FNAME` and the rest as `LNAME`. Seems like it would work in most cases and IMO is way better than having no name at all.

**Comment by zkhalapyan** [comment]
@sam-jg would work for some simple names and for small startups, but vast majority of international names would start breaking which would probably not be acceptable for medium to large size companies.

**Comment by sam-jg** [comment]
@zkhalapyan I would guess that the currently implementation is already not acceptable for medium to large size companies.
 Do you have some examples of international names that would cause issues, I don't have a lot of experience in that area sorry.

**Comment by zkhalapyan** [comment]
Checkout [Personal names around the world](https://www.w3.org/International/questions/qa-personal-names). In some countries, the order is Family-Name Given-Name.

**Comment by ksiskar** [comment]
Following + 

Any update on this? Seconding the need to have the first name and last name in Mailchimp, as every email we send starts with "Hello FNAME, "

**Comment by sam-jg** [comment]
I was hoping to get input from a maintainer before submitting a PR. It should be a pretty small change, so you could probably just submit one anyway @ksiskar.

I don't see any easy fix to the name order issue, but don't think it really matters. To me, some names in the wrong order is better than no names at all.

**Comment by russellwheatley** [comment]
Hi there! I would hold back on submitting PRs for the mailchimp extension at the moment, as far as I'm aware, there are potential changes coming so development is on hold in the meanwhile.

**Comment by eddygent** [comment]
I too am having a similar issue here, it doesn't appear FNAME and LNAME are being applied to the merge fields when making the REST call, so I'm hit with errors like:

Error when adding user to Mailchimp audience { Error: Your merge fields were invalid.
    at Request._callback (/workspace/node_modules/mailchimp-api-v3/index.js:518:35)
    at Request.self.callback (/workspace/node_modules/request/request.js:185:22)
    at Request.emit (events.js:198:13)
    at Request.EventEmitter.emit (domain.js:466:23)
    at Request.<anonymous> (/workspace/node_modules/request/request.js:1154:10)
    at Request.emit (events.js:198:13)
    at Request.EventEmitter.emit (domain.js:466:23)
    at IncomingMessage.<anonymous> (/workspace/node_modules/request/request.js:1076:12)
    at Object.onceWrapper (events.js:286:20)
    at IncomingMessage.emit (events.js:203:15)
    at IncomingMessage.EventEmitter.emit (domain.js:466:23)
    at endReadableNT (_stream_readable.js:1145:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
  type: 'https://mailchimp.com/developer/marketing/docs/errors/',
  title: 'Invalid Resource',
  status: 400,
  detail: 'Your merge fields were invalid.',
  instance: 'bb199068-472d-44bb-9eaa-80a71d61a47f',
  errors:
   [ { field: 'FNAME', message: 'Please enter a value' },
     { field: 'LNAME', message: 'Please enter a value' } ],

Look forward to seeing an update here :)

Ed

**Comment by dackers86** [comment]
Closing in favour https://github.com/mailchimp/Firebase/issues/30
